### PR TITLE
Change Breakpoints panel to use pre-syntax-highlited snippets

### PR DESCRIPTION
**Pull request by AnshVM**
Fixes issue [#8795 ](https://github.com/replayio/devtools/issues/8795) 
BEFORE:
image
AFTER: 
image

(Both are supposed to be the same)

### CODE REVIEW

File: packages/replay-next/components/sources/SyntaxHighlightedLine.tsx [hunk 1]
```diff
@@ -1,27 +1,39 @@
 import { Suspense } from "react";
 
 import { parse, parsedTokensToHtml } from "replay-next/src/suspense/SyntaxParsingCache";
+import { ParsedToken } from "replay-next/src/suspense/SyntaxParsingCache";
```
Comment (Domiii) [inline packages/replay-next/components/sources/SyntaxHighlightedLine.tsx:4]: Don't import the same file twice

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx [hunk 2]
```diff
@@ -2,9 +2,11 @@ import React, { Suspense, useContext, useMemo } from "react";
 
 import Loader from "replay-next/components/Loader";
 import SyntaxHighlightedLine from "replay-next/components/sources/SyntaxHighlightedLine";
+import { getStreamingSourceContentsSuspense } from "replay-next/src/suspense/SourcesCache";
+import { getParsedValueIfCached } from "replay-next/src/suspense/SyntaxParsingCache";
+import { ParsedToken } from "replay-next/src/suspense/SyntaxParsingCache";
```
Comment (Domiii) [inline src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx:7]: ditto: dup import

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx [hunk 3]
```diff
@@ -54,6 +69,7 @@ export default function BreakpointOptions({ breakpoint, type }: BreakpointProps)
             log(
             <SyntaxHighlightedLine code={content} />)
           </span>
+          ;
```
Comment (Domiii) [inline src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx:72]: Should this not be in `content`?

**Review by Domiii — CHANGES_REQUESTED** [review]
hi @AnshVM, and thank you for your (first?) contribution!
Two more important things to look out for:

1. Please provide a test. I know, there are not many tests in this area, but we do need to make sure that external contributions are reliable. You can consider (for example) [Focuser.test.tsx](https://github.com/replayio/devtools/blob/ebe0f3325933e39297f8285506b09d3ec5d6679d/packages/replay-next/components/console/Focuser.test.tsx).
2. Can you provide before + after screenshots, instead of only "after"? Thank you!

For more info, please refer to our guidelines [here](https://docs.replay.io/contribute/contributing-to-replay#4b7db91af691410fbb21aef607ccc15c).

File: packages/replay-next/components/sources/SyntaxHighlightedLine.tsx (hunk 1 above)
Comment (AnshVM) [reply packages/replay-next/components/sources/SyntaxHighlightedLine.tsx]: done

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx (hunk 2 above)
Comment (AnshVM) [reply src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx]: done

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx (hunk 3 above)
Comment (AnshVM) [reply src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx]: done

**Comment by AnshVM** [comment]
> hi @AnshVM, and thank you for your (first?) contribution! Two more important things to look out for:
> 
> 1. Please provide a test. I know, there are not many tests in these folders (yet!), but we do need to make sure that external contributions are reliable. You can consider (for example) [Focuser.test.tsx](https://github.com/replayio/devtools/blob/ebe0f3325933e39297f8285506b09d3ec5d6679d/packages/replay-next/components/console/Focuser.test.tsx).
> 2. Can you provide before + after screenshots, instead of only "after"? Thank you!
> 
> For more info, please refer to our guidelines [here](https://docs.replay.io/contribute/contributing-to-replay#4b7db91af691410fbb21aef607ccc15c).

Hey, thanks for the review! I have made the changes you requested.
As for the tests, I have added two tests for the SyntaxHighlightedLine component, let me know if that would be enough or if you have any suggestions for the tests.

**Comment by Domiii** [comment]
@AnshVM

> Hey, thanks for the review! I have made the changes you requested. As for the tests, I have added two tests for the SyntaxHighlightedLine component, let me know if that would be enough or if you have any suggestions for the tests.

Thanks for putting in the hard work, it is much appreciated!
One thing about testing UI or data changes: You want to avoid copy-and-pasting `html` or other application state into the test itself, since that will change over time and should be accounted for implicitely.

To better tackle this issue please employ snapshot testing:
* Remove the `html` snippet and replace it with snapshotting.
  * Ref: https://jestjs.io/docs/snapshot-testing
  * Feel free to consider samples in the code, such as [`timeline.test.tsx`](https://github.com/replayio/devtools/blob/2de522bb51101c4e836a261052f69c54ade5edfd/src/ui/reducers/timeline.test.ts)

**Comment by AnshVM** [comment]
@Domiii Thanks for the suggestion! I have made the change, let me know!

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "dev": "next dev -p 8080 ",
     "preprod": "NEXT_PUBLIC_API_URL=http://graphql-api.pre-prod.replay.prod/v1/graphql NEXT_PUBLIC_DISPATCH_URL=ws://agent.pre-prod.replay.prod next dev -p 8080",
     "export": "next build && next export -o dist",
-    "test": "jest",
+    "test": "jest -t syntax-highlighting",
```
Comment (bvaughn) [inline package.json:10]: Please revert this change.
```suggestion
    "test": "jest",
```

File: packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  it("syntax highliting should work without passing tokens to SyntaxHighlitedLine", async () => {
+    const tree = renderer.create(<SyntaxHighlightedLine code={SNIPPET} />).toJSON()
+    expect(tree).toMatchSnapshot();
+  });
+
+  it("syntax highliting should work when tokens are passed to SyntaxHighlitedLine", async () => {
+    const tree = renderer.create(<SyntaxHighlightedLine code={SNIPPET} tokens={SNIPPET_TOKENS} />).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+});
```
Comment (bvaughn) [inline packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx:43]: I understand why you're using react-test-library for this, but we have an existing convention for e2e testing in the replay-next package and I don't think we should introduce a _new_ way of writing these tests just for a single case (particularly without discussion).

We appreciate new tests! If you'd like to add one, check out the README in this directory for more information about how we write this type of test:
https://github.com/replayio/devtools/tree/main/packages/replay-next/playwright

I also think it's okay in this case to rely on existing test coverage to prevent a regression.

File: package.json [hunk 6]
```diff
@@ -200,6 +201,7 @@
     "prettier-plugin-import-sort": "^0.0.7",
     "prettier-plugin-tailwindcss": "^0.1.7",
     "raw-loader": "^4.0.2",
+    "react-test-renderer": "^18.2.0",
```
Comment (bvaughn) [inline package.json:204]: Please revert changes to `package.json` and `yarn.lock` 😄  I think taking on a new testing library dependency is something we would want to discuss and consider on its own, rather than as a side effect of a change like this.

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx [hunk 7]
```diff
@@ -17,25 +18,38 @@ function BreakpointLineContents({ breakpoint }: BreakpointProps) {
...
+      rawTextByLine &&
+      rawTextByLine.length > 0 &&
+      parsedTokensByLine &&
+      parsedTokensByLine[line - 1]
+    ) {
+      const lineText = rawTextByLine[line - 1];
+      tokens = parsedTokensByLine[line - 1].slice(
+        parsedTokensByLine[line - 1].findIndex(parsedToken => parsedToken.columnIndex === column)
+      );
+      return { snippet: lineText.slice(column, column! + 100).trim(), tokens };
```
Comment (bvaughn) [inline src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx:42]: I think these two should be checked separately?

```suggestion
const { snippet, tokens } = useMemo((): {
  snippet: string;
  tokens: ParsedToken[];
} => {
  const { column, line } = breakpoint.location;

  let snippet: string | null = null;
  let tokens: ParsedToken[] | null = null;

  if (rawTextByLine && rawTextByLine.length > 0) {
    const line = rawTextByLine[line - 1];
    snippet = line.slice(column, column! + 100).trim();
  }

  if (parsedTokensByLine && parsedTokensByLine[line - 1]) {
    const line = parsedTokensByLine[line - 1];
    tokens = line.slice(
      line.findIndex((parsedToken) => parsedToken.columnIndex === column)
    );
  }

  return { snippet, tokens };
}, [rawTextByLine, parsedTokensByLine, breakpoint]);
```

It's possible we will have loaded the plaintext source code without parsing it.

File: packages/replay-next/components/sources/SyntaxHighlightedLine.tsx [hunk 8]
```diff
@@ -1,27 +1,42 @@
...
-function SyntaxHighlightedLineSuspends({ code, className = "", fileExtension = ".js" }: Props) {
-  const tokens = parse(code, fileExtension);
-  const html = parsedTokensToHtml(tokens?.[0] ?? []);
+function SyntaxHighlightedLineSuspends({
+  code,
+  className = "",
+  fileExtension = ".js",
+  tokens,
+}: Props) {
+  tokens = tokens || (parse(code, fileExtension)?.[0] ?? []);
```
Comment (bvaughn) [inline packages/replay-next/components/sources/SyntaxHighlightedLine.tsx:38]: Kind of a nit but this is a pretty dense statement. I personally prefer to write stuff like this out a bit more explicitly, but admittedly it's a bit subjective:
```ts
if (tokens == null) {
  const parsed = parse(code, fileExtension);
  if (parsed?.length > 0) {
    tokens = parsed[0];
  }
}
```

File: packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx (hunk 5 above)
Comment (Domiii) [reply packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx:43]: Thanks @bvaughn - this is on me.
I recommended some random resources on snapshot testing earlier and pointed to some of your existing tests, but evidently I pointed in the wrong direction 👀

Comment (bvaughn) [reply packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx:43]: @Domiii Unit vs e2e tests depends on the type of change.

File: package.json (hunk 4 above)
Comment (AnshVM) [reply package.json:10]: Yeah, forgot to revert back before pushing the changes, extremely sorry for that, will make the change.

**Comment by AnshVM** [comment]
> It's been a few weeks since I left the originally comment that spawned this issue on PR #8649. Since then, our breakpoint and log point code has been rewritten to allow them to be shared with others who have access to a recording (see #8676).
> 
> As I re-read this code, I find myself wondering if it wouldn't be better to store the snippets _in GraphQL_ (or IndexedDB) along with the log point / breakpoint – like we do for Comments. That way we wouldn't have to re-fetch source and re-parse just to display them in the side nav.
> 
> That could make a nice follow up if you're interested 😄

@bvaughn 
I would definitely love to work on that! Let me know if there's anything that could help me get started with this

File: packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx (hunk 5 above)
Comment (AnshVM) [reply packages/replay-next/components/sources/SyntaxHighlitedLine.test.tsx:43]: > I understand why you're using react-test-library for this, but we have an existing convention for e2e testing in the replay-next package and I don't think we should introduce a _new_ way of writing these tests just for a single case (particularly without discussion).
> 
> We appreciate new tests! If you'd like to add one, check out the README in this directory for more information about how we write this type of test: https://github.com/replayio/devtools/tree/main/packages/replay-next/playwright
> 
> I also think it's okay in this case to rely on existing test coverage to prevent a regression.

@bvaughn got it, so I'll just remove the tests I have added for now.

File: src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx [hunk 9]
```diff
@@ -17,25 +18,37 @@ function BreakpointLineContents({ breakpoint }: BreakpointProps) {
...
-  const snippet = useMemo(() => {
+  const parsedTokensByLine = parsed?.parsedTokensByLine;
+  const rawTextByLine = parsed?.rawTextByLine;
+
+  const { snippet, tokens } = useMemo((): { snippet: string; tokens: ParsedToken[] } => {
     const { column, line } = breakpoint.location;
     let snippet = "";
+    let tokens: ParsedToken[] = [];
+
+    if (rawTextByLine && rawTextByLine.length > 0) {
```
Comment (bvaughn) [inline src/devtools/client/debugger/src/components/SecondaryPanes/Breakpoints/BreakpointOptions.tsx:32]: If `rawTextByLine` has _some_ lines but not as many as the specified index (`line - 1`) this check will cause a runtime error below– because we'll end up calling `.slice` on undefined.

**Comment by AnshVM** [comment]
@bvaughn made the changes. Can you review them?

**Comment by bvaughn** [comment]
It's in my queue but it may take a couple of days.

**Comment by AnshVM** [comment]
> It's in my queue but it may take a couple of days.

Hey, any idea how much more time this could take? Sorry for being a little impatient here, but it would be great if I could make some progress on this before applying for GSOC. Thanks!

**Comment by bvaughn** [comment]
I understand, although reviewing these PRs aren't at the top of my priorities list.

Unfortunately it looks like there are some conflicts too, from the recent refactoring related to the "suspense" package adoption.

If you'd like to update and resolve those conflicts– ping me and I can try to take a look on Monday. If you'd rather not, that's understandable as well.

**Comment by AnshVM** [comment]
> I understand, although reviewing these PRs aren't at the top of my priorities list.
> 
> Unfortunately it looks like there are some conflicts too, from the recent refactoring related to the "suspense" package adoption.
> 
> If you'd like to update and resolve those conflicts– ping me and I can try to take a look on Monday. If you'd rather not, that's understandable as well.

@bvaughn 
I would surely like to update and resolve these conflicts.

**Comment by bvaughn** [comment]
Go for it.

**Review by bvaughn — APPROVED** [review]

**Comment by AnshVM** [comment]
> I think this can be merged now, but I'm going to wait for CI to run to double check.

Awesome,thanks for all the feedback and reviews. 
Let me know if there's some issue you think I could help with :)
